scene/3d/collision_object_3d.cpp

```diff
@@ -178,50 +186,47 @@ void CollisionObject3D::_update_debug_shapes() {
 			ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
 			for (int i = 0; i < shapedata.shapes.size(); i++) {
 				ShapeData::ShapeBase &s = shapes[i];
-				if (s.debug_shape) {
-					s.debug_shape->queue_delete();
-					s.debug_shape = nullptr;
-					--debug_shape_count;
+				if (s.debug_shape.is_null()) {
+					s.debug_shape = RS::get_singleton()->instance_create();
+					RS::get_singleton()->instance_set_scenario(s.debug_shape, get_world_3d()->get_scenario());
+					++debug_shapes_count;
 				}
 				if (s.shape.is_null() || shapedata.disabled) {
+					RS::get_singleton()->free(s.debug_shape);
+					s.debug_shape = RID();
+					--debug_shapes_count;
 					continue;
 				}
-
 				Ref<Mesh> mesh = s.shape->get_debug_mesh();
-				MeshInstance3D *mi = memnew(MeshInstance3D);
-				mi->set_transform(shapedata.xform);
-				mi->set_mesh(mesh);
-				add_child(mi);
-				mi->force_update_transform();
-				s.debug_shape = mi;
-				++debug_shape_count;
+				RS::get_singleton()->instance_set_base(s.debug_shape, mesh->get_rid());
+				RS::get_singleton()->instance_set_transform(s.debug_shape, get_global_transform() * shapedata.xform);
 			}
 		}
 	}
 	debug_shapes_to_update.clear();
 }
 
 void CollisionObject3D::_clear_debug_shapes() {
-	for (Map<uint32_t, ShapeData>::Element *E = shapes.front(); E; E = E->next()) {
-		ShapeData &shapedata = E->get();
-		ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
-		for (int i = 0; i < shapedata.shapes.size(); i++) {
-			ShapeData::ShapeBase &s = shapes[i];
-			if (s.debug_shape) {
-				s.debug_shape->queue_delete();
-				s.debug_shape = nullptr;
-				--debug_shape_count;
+	if (debug_shapes_count > 0) {
```

Comment:
Why do you need this condition here?

Reply:
No specific reason, it was here because in older versions of this PR there are 2 _clear_debug_shapes calls.